asn1ate/parser.py

```diff
@@ -94,6 +94,48 @@ def annotation(t):
 
         return annotation
 
+    def auto_tag(t):
+        tag_number = 0
+        # First search for any existing tags
+        # If there are tags then auto-tagging does not apply
+        components = t.asList()[0].elements[1]
+        for component in components:
+            if component.ty == "ComponentType":
+                elements = component.elements[0].elements # Sequence and set types
+            else:
+                elements = component.elements # Choice type
+            for element in elements:
+                if element.ty == "Type":
+                    for e in element.elements:
+                        if e.ty == "TaggedType":
+                            return
+        
+        for component in components:
+            # Convert each one into a tagged type
+            if component.ty == "ComponentType":
+                elements = component.elements[0].elements # Sequence and set types
+            else:
+                elements = component.elements # Choice type
+            for element in elements:
+                if element.ty == "Type":
+                    for e, idx in zip(element.elements, range(len(element.elements))):
+                        tagged_element = AnnotatedToken("TaggedType", [AnnotatedToken("Tag", [
```

Comment:
Ugh, the manual building of parse-tree hurts my eyes... And it seems quite sensitive if we tweak the parser. I wonder if it would be clearer/more maintainable if done in sema? Since you're doing this for sequence-type, set-type and choice-type, you should be able to do something in sema.ConstructedType, right?